libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp

```diff
@@ -80,25 +87,43 @@ void AP_RangeFinder_BLPing::update(void)
 
     const uint32_t now = AP_HAL::millis();
 
-    if (get_reading(state.distance_cm)) {
+    if (check_new_messages() && sensor_initialized) {
         // update range_valid state based on distance measured
         state.last_reading_ms = now;
         update_status();
-    } else if (now - state.last_reading_ms > BLPING_TIMEOUT_MS) {
+
+        // request new distance
+        request_message(BLDPIN_MSGID_DISTANCE_SIMPLE);
+
+        return;
+    }
+
+    if (now - state.last_reading_ms > BLPING_TIMEOUT_MS) {
         set_status(RangeFinder::RangeFinder_NoData);
 
         // initialise sensor if no distances recently
         if (now - last_init_ms > BLPING_INIT_RATE_MS) {
             last_init_ms = now;
+            // set initial states if sensor was already initialized
+            if(sensor_initialized) {
+                sensor_initialized = false;
+                ping_interval = 0;
+            }
             init_sensor();
         }
     }
 }
 
 void AP_RangeFinder_BLPing::init_sensor()
 {
-    // request distance from sensor
-    send_message(BLDPIN_MSGID_DISTANCE_SIMPLE, nullptr, 0);
+    // It's necessary to set the correct interval between
+    // pings before requesting for new distances
+    if(ping_interval != BLPING_INTERVAL_MS) {
```

Comment:
We must request distance simple every time this is called (because the timeout to trigger init depends on distance simple reception). We should always do both (set ping interval and request distance simple, no if/else), and we should do them one at a time (synchronously).

This function should probably return a bool.